### PR TITLE
feat(interaction): crosshair

### LINKED ISSUE #58: Tooltip with multiple values on hover
When multiple datapoints of a line or area chart have the same x and y value, the tooltip is displayed only for the series that reside on the top-most z-index.
The idea is to get all the possible values and display a tooltip with all the matching values

**Kibana Cross Issues**
- https://github.com/elastic/kibana/issues/30932
- https://github.com/elastic/kibana/issues/4885

**Checklist**
- [x] every related Kibana issue is listed under `Kibana Cross Issues` list
- [x] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### LINKED ISSUE #80: Crosshair and vertical/horizontal cursors
**Is your feature request related to a problem? Please describe.**
Add the possibility to switch between simple element tooltip to multi/element tooltips like vertical or horizontal cursors or crosshair.
screenshot 2019-02-28 at 15 24 02
**Describe the solution you'd like**

The user/developer can select the type of tooltip that's better conforms for the chart use to enable the following possibilities:
- single tooltip with single value
- single tooltip with multiple values (with all data thats belongs to the same data point or to the same X or Y axis value, eg stacked charts see: https://github.com/elastic/elastic-charts/issues/68)
image
See also https://www.chartjs.org/docs/latest/general/interactions/modes.html as a reference

**Describe alternatives you've considered**

Various solutions can be applied:
- remove the existing listeners to graphic elements and use directly mouse coordinates to mapping series data values getting series values from inverted x and y scales or use some spatial index for faster searches like https://github.com/mourner/flatbush
- draw a hidden layer of grouped elements with listener attached to it (can be slow and can only works on vertical and horizontal cursors.

 **design note**: the tooltip could be crowded specially with multiple splitted series.
**extra**: this can solve the following issue: https://github.com/elastic/elastic-charts/issues/58

**Additional context**

TVSB, charts used by Stack Monitor and any other chart that use react-vis in kibana currently has the possibility to include a vertical cursor on the chart instead of the tooltip.

**Kibana Cross Issues**
- [ ] https://github.com/elastic/kibana/pull/14914
- [ ] https://github.com/elastic/kibana/issues/13864

**Checklist**

- [x] this request is checked against already exist requests
- [ ] every related Kibana issue is listed under `Kibana Cross Issues` list
- [x] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### LINKED ISSUE #88: Tooltip x value field
**Describe the bug**
The tooltip label the x value as `X Value` if not title on horizontal axis is available.
If multiple horizontal axis are available and only one axis title configured, the tooltip will shows `X Value` also if we have titles on it.

**To Reproduce**
1. Go to Area Chart - with 4 axes storybook story (remove a title from one of the two horizontal axes)
2. Hover over any data point

**Expected behavior**
Always display at least one title independently of the status of the axis titles.

We also have to include the possibility to add an hidden title on an axis so we can have the label on the tooltip but no axis title displayed on the chart

**Screenshots**
screenshot 2019-03-01 at 16 57 14

**Version (please complete the following information):**
 - OS: osx
 - Browser: chrome
 - Elastic Charts: > 2.0.0

**Additional context**
n/a

**Errors in browser console**
n/a

**Kibana Cross Issues**
n/a

**Checklist**
- [x] every related Kibana issue is listed under `Kibana Cross Issues` list
- [x] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### PR DESCRIPTION
**Pull request by markov00**
close #80 
close #58
close #88

## Summary
WIP

This PR introduce the Crosshair interaction.
Screenshot 2019-03-18 at 19 23 07

We have removed any listeners to canvas elements. We are instead using a map for each value on the chart: this map maps x value and any relative Y values associated. We then just invert the x scale relative to the mouse point to get the right X key to use on that map. This increase the performance on highly/medium dense charts.

The inverted function on time scales, currently works correctly when using UTC dates. Further investigation needs to be done to make that works with any time zone. To test this check `Interactions - brush selection tool on time charts` story and remove the `Z` from the `now` constant time.

WIP

ToDo:
- [x] add crosshair style to theme
- [x] adjust crosshair tooltip position
- [x] fix tooltip style 
    - [x] fix tooltip position on rotated charts
    - [x] highlight element/s
         - [x] add increased radius point on line/area chart
         - [x] add highlighted bar on barchart
         - [x] add background on whole stack/cluster group
         - [x] add highlight on tooltip for hovered elements
    - [x] add colors beside each series
    - [x] fix series labels with the right series name
    - [x] hide hidden series from legend
- [x] invert crosshair on rotated charts
- [x] add crosshair props to spec
- [x] convert/ revert element on hover handlers
- [x] style a bit better the highlighted elements
- [ ] check area/line legend on hover style

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Proper documentation or storybook story was added for features that require explanation or tutorials
- [ ] Unit tests were updated or added to match the most common scenarios
- [ ] Each commit follows the [convention](https://github.com/elastic/elastic-charts/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/state/chart_state.ts [hunk 1]
```diff
@@ -171,54 +199,190 @@ export class ChartStore {
     this.legendCollapsed.set(!this.legendCollapsed.get());
     this.computeChart();
   });
-  onOverElement = action((tooltip: TooltipData) => {
-    if (this.onElementOverListener) {
-      this.onElementOverListener(tooltip.value);
+
+  setCursorPosition = action((x: number, y: number) => {
```
Comment (emmacunningham) [inline src/state/chart_state.ts]: Perhaps some of the computation in this function can be pulled into utils as pure functions which (hopefully) will make test writing a bit easier.

File: src/state/chart_state.ts [hunk 2]
```diff
@@ -171,54 +199,190 @@ export class ChartStore {
...
+    // if no element, hide everything
+    if (!elements || elements.length === 0) {
+      this.clearTooltipAndHighlighted();
+      return;
+    }
+
+    // build the tooltip value list
+    let xValueInfo: TooltipValue | null = null;
+    let oneHighlighted = false;
+    const newHighlightedGeometris: IndexedGeometry[] = [];
```
Comment (emmacunningham) [inline src/state/chart_state.ts]: minor typo :)

```suggestion
    const newHighlightedGeometries: IndexedGeometry[] = [];
```

File: src/state/crosshair_utils.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+export function getSnapPosition(value: string | number, scale: Scale, totalGroupCount: number = 1) {
+  if (isOrdinalScale(scale)) {
+    return getOrdinalSnapPosition(value, scale, totalGroupCount);
+  } else {
+    return getContinuousSnapPosition(value as number, scale as ScaleContinuous, totalGroupCount);
+  }
+}
+
+function getOrdinalSnapPosition(value: string | number, scale: ScaleBand, totalGroupCount: number) {
```
Comment (emmacunningham) [inline src/state/crosshair_utils.ts]: for this and `getContinuousSnapPosition`, could we create an interface for the return value?

File: src/components/crosshair.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      chartRotation,
+      tooltipType,
+      chartTheme: { crosshair },
+    } = this.props.chartStore!;
+    if (
+      !isCrosshairTooltipType(tooltipType.get()) ||
+      cursorPosition.x === -1 ||
+      cursorPosition.y === -1
+    ) {
+      return <div className="elasticChartsCrosshair" />;
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/crosshair.tsx]: Nothing really special, it just avoid recreating a the container DOM node after the diff. 1 less DOM node to create, not really a performance booster :)

File: src/state/crosshair_utils.ts (hunk 3 above)
Comment (markov00) [reply src/state/crosshair_utils.ts]: resolved on https://github.com/elastic/elastic-charts/pull/96/commits/7212d5121d35f43273c6a3eb490d160d5844f928

File: src/state/chart_state.ts (hunk 1 above)
Comment (markov00) [reply src/state/chart_state.ts]: partially solved https://github.com/elastic/elastic-charts/pull/96/commits/7212d5121d35f43273c6a3eb490d160d5844f928

File: src/components/tooltips.tsx [hunk 5]
```diff
@@ -1,58 +1,81 @@
...
+        vPosition = {
+          position: 'top',
+          value: cursorPosition.y,
+        };
+      } else {
+        vPosition = {
+          position: 'bottom',
+          value: chartDimensions.height - cursorPosition.y,
+        };
+      }
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/tooltips.tsx]: solved https://github.com/elastic/elastic-charts/pull/96/commits/7212d5121d35f43273c6a3eb490d160d5844f928

File: src/components/tooltips.tsx [hunk 6]
```diff
@@ -1,58 +1,81 @@
...
     let hPosition;
-    if (!initialized.get() || !tooltip || !tooltipPos) {
+    if (
+      !initialized.get() ||
+      tooltipType.get() === TooltipType.None ||
+      !tooltipData ||
+      tooltipData.length === 0 ||
+      cursorBandPosition.x === -1 ||
+      cursorBandPosition.y === -1 ||
+      cursorPosition.y === -1
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/tooltips.tsx]: solved https://github.com/elastic/elastic-charts/pull/96/commits/7212d5121d35f43273c6a3eb490d160d5844f928

File: src/state/chart_state.ts [hunk 7]
```diff
@@ -79,28 +90,28 @@ export interface SeriesDomainsAndData {
...
-export type ElementOverListener = (value: GeometryValue) => void;
+export type ElementClickListener = (values: GeometryValue[]) => void;
+export type ElementOverListener = (values: GeometryValue[]) => void;
 export type BrushEndListener = (min: number, max: number) => void;
 export type LegendItemListener = (dataSeriesIdentifiers: DataSeriesColorsValues | null) => void;
 // const MAX_ANIMATABLE_GLYPHS = 500;
 
 export class ChartStore {
-  debug = false;
+  debug = true;
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/state/chart_state.ts]: solved in https://github.com/elastic/elastic-charts/pull/96/commits/7212d5121d35f43273c6a3eb490d160d5844f928

File: src/components/crosshair.tsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    return (
+      <div className="elasticChartsCrosshair">
+        {canRenderVertical(tooltipType.get(), chartRotation, crosshair.vertical.visible) &&
+          this.renderVertical()}
+        {canRenderHorizontal(tooltipType.get(), crosshair.horizontal.visible) &&
+          this.renderHorizontal()}
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/crosshair.tsx]: solved https://github.com/elastic/elastic-charts/pull/96/commits/7212d5121d35f43273c6a3eb490d160d5844f928

File: src/components/crosshair.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      left,
+      width,
+      top,
+      height: 0,
+      borderTopWidth: crossHairStyle.strokeWidth,
+      borderTopColor: crossHairStyle.stroke,
+      borderTopStyle: crossHairStyle.dash ? 'dashed' : 'solid',
+      background: 'transparent',
+      zIndex: 190,
+      opacity: crossHairStyle.opacity,
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/crosshair.tsx]: solved https://github.com/elastic/elastic-charts/pull/96/commits/7212d5121d35f43273c6a3eb490d160d5844f928

File: src/state/chart_state.ts (hunk 2 above)
Comment (markov00) [reply src/state/chart_state.ts]: resolved in https://github.com/elastic/elastic-charts/pull/96/commits/8a0bcb157c258f78adc1b5476713d24d4002400f

File: src/lib/series/series_utils.ts [hunk 10]
```diff
@@ -27,3 +31,11 @@ export function belongsToDataSeries(geometryValue: GeometryId, dataSeriesValues:
 
   return hasSameSpecId && hasSameSeriesKey;
 }
+
+export function getYAccessor(spec: BasicSeriesSpec, seriesKey: any[]) {
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/lib/series/series_utils.ts]: removed in https://github.com/elastic/elastic-charts/pull/96/commits/8a0bcb157c258f78adc1b5476713d24d4002400f

File: src/components/react_canvas/area_geometries.tsx [hunk 11]
```diff
@@ -93,57 +53,97 @@ export class AreaGeometries extends React.PureComponent<
...
-            stroke={stroke}
             strokeWidth={strokeWidth}
-            opacity={overPoint === areaPoint ? 1 : 0}
+            stroke={stroke}
+            fill={color}
+            opacity={opacity}
             strokeHitEnabled={false}
             listening={false}
-            perfectDrawEnabled={true}
+            perfectDrawEnabled={false}
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/components/react_canvas/area_geometries.tsx]: will work on that on a different PR

**Pull request merged**